src/aes128gcm.rs

```diff
@@ -23,174 +28,380 @@ const ECE_WEBPUSH_IKM_LENGTH: usize = 32;
 const ECE_AES128GCM_KEY_INFO: &str = "Content-Encoding: aes128gcm\0";
 const ECE_AES128GCM_NONCE_INFO: &str = "Content-Encoding: nonce\0";
 
-// TODO: When done, remove the aes128gcm prefixes and the EC_ ones.
-// As for now it makes it easier to Ctrl + F into ecec :)
+/// Encrypts a Web Push message using the "aes128gcm" scheme, with an explicit sender key.
+///
+/// It is the caller's responsibility to ensure that this function is used correctly,
+/// where "correctly" means important cryptographic details like:
+///
+///    * use a new ephemeral local keypair for each encryption
+///    * use a randomly-generated salt
+///
+/// In general-purpose AES128GM ECE, the "keyid" field in the header may be up to 255 octects
+/// and provides a string that allows the application to find the right key material in some
+/// application-defined way. We only currently support the specific scheme used by WebPush, where
+/// the "keyid" is an ephemeral ECDH public key and always has a fixed length.
+///
+pub(crate) fn encrypt(
+    local_prv_key: &dyn LocalKeyPair,
+    remote_pub_key: &dyn RemotePublicKey,
+    auth_secret: &[u8],
+    plaintext: &[u8],
+    mut params: WebPushParams,
+) -> Result<Vec<u8>> {
+    let cryptographer = crypto::holder::get_cryptographer();
+
+    if plaintext.is_empty() {
+        return Err(Error::ZeroPlaintext);
+    }
+
+    let salt = params.take_or_generate_salt(cryptographer)?;
+    let (key, nonce) = derive_key_and_nonce(
+        cryptographer,
+        EceMode::ENCRYPT,
+        local_prv_key,
+        remote_pub_key,
+        auth_secret,
+        &salt,
+    )?;
+
+    // Encode the ephemeral public key in the "kid" header field.
+    let keyid = local_prv_key.pub_as_raw()?;
+    if keyid.len() != ECE_WEBPUSH_PUBLIC_KEY_LENGTH {
+        return Err(Error::InvalidKeyLength);
+    }
+
+    let header = Header {
+        salt: &salt,
+        rs: params.rs,
+        keyid: &keyid,
+    };
+
+    // We always add at least one padding byte, for the delimiter.
+    let padding = std::cmp::max(params.pad_length, ECE_AES128GCM_PAD_SIZE);
 
-/// Web Push encryption structure for the AES128GCM encoding scheme ([RFC8591](https://tools.ietf.org/html/rfc8291))
+    // For now, everything must fit in a single record.
+    // Calling code will ensure that this is the case.
+    if params.rs < ECE_AES128GCM_MIN_RS {
+        return Err(Error::InvalidRecordSize);
+    }
+    if plaintext.len() + padding + ECE_TAG_LENGTH > params.rs as usize {
```

Comment:
In *theory* we should never hit this branch, because the calling code set `rs` appropriate; but in practice, future-you or future-me will probably need such a debugging message at some point 😅 - thanks, I'll add one.